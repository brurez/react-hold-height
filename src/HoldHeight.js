import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

class HoldHeight extends Component {
  constructor(props) {
    super(props);
    this.state = {
      height: props.initialHeight || 0,
      windowWidth: 0,
      reset: !(props.resetOnWindowResize === false)
    };
    this.resetHeight = _.debounce(this.resetHeight, 300);
  }

  componentDidMount() {
    window.addEventListener('resize', this.resetHeight.bind(this));
    this.updateHeight();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.resetHeight.bind(this));
  }

  resetHeight() {
    if (
      this.state.windowWidth !== window.innerWidth &&
      this.state.reset === true
    ) {
      this.setState({
        height: this.props.initialHeight || 0,
        windowWidth: window.innerWidth
      });
    }
  }

  componentWillReceiveProps(next) {
    if (!_.isEqual(next.children, this.props.children)) {
      this.updateHeight();
    }
  }

  updateHeight() {
    let stateHeight = _.clone(this.state.height);
    const height = this.me.clientHeight;

    if (height > stateHeight) this.setState({ height });
  }

  render() {
    return (
      <div
        ref={me => {
          this.me = me;
        }}
        style={{ minHeight: this.state.height }}
      >
        {this.props.children}
      </div>
    );
  }
}

HoldHeight.propTypes = {
  resetOnWindowResize: PropTypes.bool,
  initialHeight: PropTypes.number
};

export default HoldHeight;
