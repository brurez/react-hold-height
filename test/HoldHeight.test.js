import React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

import HoldHeight from '../src/index';

describe('<HoldHeight />', () => {

  const List3Items = () =>
    <ul>
      <li>This is a list</li>
      <li>with 3 itens</li>
      <li>it takes some space</li>
    </ul>;

  it('render child with no props', () => {
    const wrapper = mount(<HoldHeight><List3Items/></HoldHeight>);
    expect(wrapper.find('li')).toHaveLength(3);
    expect(wrapper).toMatchSnapshot();
  });

  it('render child with initialHeight prop', () => {
    const wrapper = mount(<HoldHeight initialHeight={200}><List3Items/></HoldHeight>);
    expect(wrapper.find('li')).toHaveLength(3);
    expect(wrapper.props().initialHeight).toBe(200);
    expect(wrapper).toMatchSnapshot();
  });

});
