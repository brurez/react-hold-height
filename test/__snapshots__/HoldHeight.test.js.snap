// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HoldHeight /> render child with initialHeight prop 1`] = `
<HoldHeight
  initialHeight={200}
>
  <div
    style={
      Object {
        "minHeight": 200,
      }
    }
  >
    <List3Items>
      <ul>
        <li>
          This is a list
        </li>
        <li>
          with 3 itens
        </li>
        <li>
          it takes some space
        </li>
      </ul>
    </List3Items>
  </div>
</HoldHeight>
`;

exports[`<HoldHeight /> render child with no props 1`] = `
<HoldHeight>
  <div
    style={
      Object {
        "minHeight": 0,
      }
    }
  >
    <List3Items>
      <ul>
        <li>
          This is a list
        </li>
        <li>
          with 3 itens
        </li>
        <li>
          it takes some space
        </li>
      </ul>
    </List3Items>
  </div>
</HoldHeight>
`;
